### DISCUSSION TITLE
ImagePolicy with Buckets

### DISCUSSION THREAD

**Discussion by BDworak** [fluxcd/flux2#3479] [Q&A] [answered]
Hello,

I am trying to setup flux to sync and apply buckets from an Azure Storage account. One item I would like to enable is automatic image version bumping. Currently we release versioned software to our ACR in the format of 1.x. 

I created a bucket yaml pointing to an Azure storage account and was able to successfully sync said bucket. 
I created a flux Kustomization yaml, referencing the bucket + path to apply and am able to successfully apply yamls from said bucket on a scheduled interval.
I created an ImageRepository yaml pointing at a specific image in our ACR and the tag sync worked successfully. 
I created an ImagePolicy yaml and it is able to point out the newest version of our container image based on the filters supplied (version is 1.214)

When I upload to our bucket a deployment yaml referencing an old image, image automation does not automatically replace this image to the newer version as discovered by ImagePolicy. 

I tried to add the # {"$imagepolicy":  comment to our yamls stored inside of the bucket but this did not work either. 

Is this something that is supported in fluxV2? In my environment, pushing back to git is not simple and has a lot of red-tape, so storing old image references in Git would be ideal and having an ImagePolicy apply the newest image(s) when it changes would be ideal if possible.

**Comment by kingdonb** [comment, marked as answer]
ImageUpdateAutomation only works with GitRepository sources. If you are trying to get your ImageUpdateAutomation to produce changes in a Bucket source, that will definitely not work.

I think I understand the root issue, that you want to avoid writing back to git from CI because it's a security problem, since Git is the source of truth, if CI can write back to git, then potentially untested code can potentially write back to git; there are mitigations you can apply for this, but it's much better if you just don't have to write back to Git at all, no permissions to manage at all.

We are seeing adopters find solutions that help them obviate the use of Image Update Automations in many cases (which can only write back to git) through the use of the new OCIRepository feature. Many of the possibilities here are not yet fully explored, and so I do not have a doc I can point at that explains what I mean exactly, but here's a link to the OCI Cheatsheet which covers some basics:

https://fluxcd.io/flux/cheatsheets/oci-artifacts/

What I'd be suggesting is that, instead of having Flux look for new images and try to write them back to a source of truth, let your CI release process take ownership of "advertising" these artifacts. When you release a new version, change the version number in the pipeline, in a non-main branch with automation, or as part of the release PR (someone who manages the release updates Chart.yaml for example, with the new version to be released – they can make this change in other places like YAML manifests at the same time)

Then, publish those artifacts as OCIRepository, and use one of the OCIRepository `spec.ref` behaviors like `semver` to manage your upgrades, or use ImageUpdateAutomation there.

It is difficult to offer a solution that does not write back to Git, there is definitely a missing doc or two here. We need to have some additional thought about how to solve this. But hopefully this gives you some ideas!

**Reply by BDworak** [reply]
Thank you for the quick, thorough, and direct response. I will look into the OCI artifacts and see if this satisfies my needs.
